### DISCUSSION TITLE
Make production build more quiet?

### DISCUSSION THREAD

**Discussion by m1ga** [appcelerator/titanium_mobile#12816] [Q&A] [answered]
_first: yes it is very easy/quick to check if it is an Appcelerator app. eg. use Dexplorer and look at the AndroidManifset._

Should the production build not log any Ti.API.* infos? E.g. there is always a ["powered by" message](https://github.com/appcelerator/titanium_mobile/blob/fe3deddcea07ecb3555673cf2dba8d676fd79611/common/Resources/ti.main.js#L19). Android will only hide Ti.API.debug in production builds so it is always visible if you don't remove it on your own. I think maybe it shouldn't log anything there, if you want to get some logs: create a debug version :smile: 
There was [ti-stealth](https://www.tidev.io/2014/04/17/remove-log-calls-from-production-code/) or other workarounds to create a custom log function but for security reasons a store build should not contain any log output that can be easily spotted with `adb logcat`

Sources for the logging inside the SDK: https://github.com/appcelerator/titanium_mobile/blob/master/android/runtime/v8/src/native/modules/APIModule.cpp

**Comment by jquick-axway** [comment, marked as answer]
Yes, this was intentional for production builds. Just like how the native logging functions still work for release builds.

If you want to disable **all** logging functions for production/release builds, then add a new script file to your project named `DisableLogging.bootstrap.js` with the below code. This will replace all logging functions to make them no-op. Note that files ending with `*.bootstrap.js` will be loaded by Titanium before the `app.js` or `alloy.js` (it's an undocumented feature). Also note that this will not block the "Powered by Titanium" log message since bootstrap scripts are loaded afterwards, but this was the intent on our end.

```javascript
// Disable logging for release/production builds.
if (DIST_STORE) {
	const emptyFunction = () => {};
	Ti.API.info = emptyFunction;
	Ti.API.warn = emptyFunction;
	Ti.API.error = emptyFunction;
	Ti.API.debug = emptyFunction;
	Ti.API.trace = emptyFunction;
	Ti.API.timestamp = emptyFunction;
	Ti.API.log = emptyFunction;
	console.log = emptyFunction;
	console.warn = emptyFunction;
	console.error = emptyFunction;
	console.debug = emptyFunction;
	console.trace = emptyFunction;
}
```

Does this help?

**Reply by m1ga** [reply]
oh, thats a nice feature 👍  That will help a lot and is a very clean solution. Thanks

**Reply by jquick-axway** [reply]
Happy to help!

FYI: Titanium loads all bootstrap scripts on app startup here...
https://github.com/appcelerator/titanium_mobile/blob/master/common/Resources/ti.main.js#L31

The CLI generates a `bootstrap.json` during the build which provides a list of all files ending with `.bootstrap.js` in the app project's file tree... including modules. This JSON helps with fast startup times.

We use this feature with our `ti.playservices` module. It adds a bootstrap script which checks if Google Play Services is installed and using the expected version on startup. If not installed/updated, then it will show a dialog asking the end-user to install/update it before procedding to load the "app.js".
https://github.com/appcelerator-modules/ti.playservices/blob/master/android/Resources/ti.playservices/ti.playservices.bootstrap.js
